src/services/blocks/trace/Trace.ts

```diff
@@ -0,0 +1,690 @@
+	private static getKeyNames(api: ApiPromise): Record<string, KeyInfo> {
+		const moduleKeys = Object.keys(api.query).reduce((acc, mod) => {
+			Object.keys(api.query[mod]).forEach((item) => {
+				const queryObj = api.query[mod][item];
+				let key;
+				try {
+					// `slice(2)` is to slice off '0x' prefix
+					key = queryObj.key().slice(2);
+				} catch {
+					key = queryObj.keyPrefix().slice(2);
+				}
+
+				acc[key.toString()] = {
+					module: mod,
+					item,
+				};
+			});
+
+			return acc;
+		}, {} as Record<string, KeyInfo>);
+
+		return { ...moduleKeys, ...WELL_KNOWN_KEYS };
+	}
+
+	/**
+	 * Extract extrinsic indexes from all `:extrinsic_index` storage item read events.
+	 *
+	 * Note: Expects the given `spanIds` are from an apply extrinsic action group because
+	 * apply extrinsic action groups should contain read events to `:extrinsic_index`.
+	 *
+	 * @param spanIds List of spanIds for an apply extrinsic action group
+	 * @param extrinsicIndexBySpanId Map of spanId => extrinsicIndex
+	 */
+	private static extractExtrinsicIndex(
+		spanIds: number[],
+		extrinsicIndexBySpanId: ExtrinsicIndexBySpanId
+	): BN {
+		// There are typically multiple reads of extrinsic index when applying an extrinsic,
+		// So we get all of those reads, which should all be the same number.
+		const extrinsicIndex = spanIds.reduce((acc, id) => {
+			if (!extrinsicIndexBySpanId.has(id)) {
+				return acc;
+			}
+
+			acc.push(extrinsicIndexBySpanId.get(id) as BN);
+			return acc;
+		}, [] as BN[]);
+
+		if (extrinsicIndex.length === 0) {
+			// Since we assume the `spanIds` are for spans from an apply extrinsic action group,
+			// we assume there is always an `:extrinsic_index` event
+			throw new InternalServerError('Expected at least one extrinsic index');
+		}
+
+		if (extrinsicIndex.length > 1) {
+			// If there are multiple reads we check that they are all the same. If they
+			// are not this program has some incorrect assumptions and this error should
+			// be reported to the maintainers.
+			for (const [i, extIdx] of extrinsicIndex.entries()) {
+				if (i > 0 && !extrinsicIndex[i - 1]?.eq(extIdx)) {
+					throw new InternalServerError(
+						'Expect extrinsic to only be applied at a single index.'
+					);
+				}
+			}
+		}
+
+		return extrinsicIndex[0];
+	}
+
+	/**
+	 * Find the Ids of all the spans which are descendant of the span `root`.
+	 *
+	 * @param root span which we want all the descendants of
+	 * @param spansById map of span id => `SpanWithChildren`
+	 */
+	private static findDescendants(root: number, spansById: SpansById): number[] {
+		// Note: traversal order doesn't matter here
+		const stack = spansById.get(root)?.children;
+		if (!stack) {
+			return [];
+		}
+
+		const descendants = [];
+		while (stack.length) {
+			const curId = stack.pop();
+			if (!curId) {
+				break;
+			}
+
+			descendants.push(curId);
+
+			const curSpanChildren = spansById.get(curId)?.children;
+			curSpanChildren && stack.push(...curSpanChildren);
```

Comment:
this syntax looks weird to me, is this an idiom in TS/JS? 

Personally I found an if, much more readable

```
 if (curSpan) {
    stack.push(...curSpanChildren);
 }
```